admin/sql/CreateFunctions.sql

```diff
@@ -1490,6 +1490,101 @@ CREATE OR REPLACE FUNCTION controlled_for_whitespace(TEXT) RETURNS boolean AS $$
   SELECT NOT padded_by_whitespace($1) AND whitespace_collapsed($1);
 $$ LANGUAGE SQL IMMUTABLE SET search_path = musicbrainz, public;
 
+CREATE OR REPLACE FUNCTION update_aggregate_tag_count(entity_type taggable_entity_type, entity_id INTEGER, tag_id INTEGER, count_offset SMALLINT)
+RETURNS VOID AS $$
+BEGIN
+  -- Insert-or-update the aggregate vote count for the given (entity_id, tag_id).
+  EXECUTE format(
+    $SQL$
+      INSERT INTO %1$I AS agg (%2$I, tag, count)
+           VALUES ($1, $2, $3)
+      ON CONFLICT (%2$I, tag) DO UPDATE SET count = agg.count + $3
+    $SQL$,
+    entity_type::TEXT || '_tag',
+    entity_type::TEXT
+  ) USING entity_id, tag_id, count_offset;
+END;
+$$ LANGUAGE 'plpgsql';
+
+CREATE OR REPLACE FUNCTION delete_unused_aggregate_tag(entity_type taggable_entity_type, entity_id INTEGER, tag_id INTEGER)
+RETURNS VOID AS $$
+BEGIN
+  -- Delete the aggregate tag row for (entity_id, tag_id) if no raw tag pair
+  -- exists for the same.
+  --
+  -- Note that an aggregate vote count of 0 doesn't imply there are no raw
+  -- tags; it's a sum of all the votes, so it can also mean that there's a
+  -- downvote for every upvote.
+  EXECUTE format(
+    $SQL$
+      DELETE FROM %1$I
+            WHERE %2$I = $1
+              AND tag = $2
+              AND NOT EXISTS (SELECT 1 FROM %3$I WHERE %2$I = $1 AND tag = $2)
+    $SQL$,
+    entity_type::TEXT || '_tag',
+    entity_type::TEXT,
+    entity_type::TEXT || '_tag_raw'
+  ) USING entity_id, tag_id;
+END;
+$$ LANGUAGE 'plpgsql';
+
+CREATE OR REPLACE FUNCTION update_tag_counts_for_raw_insert()
+RETURNS trigger AS $$
+DECLARE
+  entity_type taggable_entity_type;
+  new_entity_id INTEGER;
```

Comment:
There are two kinds of parameters: normal function arguments, which are declared in the `CREATE FUNCTION` line, and trigger function arguments which are passed in the `CREATE TRIGGER` command.  The latter appear in `TG_ARGV`, which is a special variable available inside trigger functions.  `TG_ARGV` is of data type `TEXT[]` though, so you can only pass strings this way.

The `DECLARE` section is just for all local variables used in the function/block below.  Like `let foo;` in JS, except they all have to go in one place at the top of the function.  I'm declaring an `entity_type` variable to hold `TG_ARGV[0]` casted to the correct type.  (So this would throw an error if the string we pass isn't a `taggable_entity_type`, which is what we want.)